### PR TITLE
Feature download parkings

### PR DESCRIPTION
**Pull request by matttsss**
## What is this PR?
This PR aims to enable the download and management of parkings in zones for offline mode.
It adds two new functions to the `ParkingViewModel`: 
* `downloadZone(...)` that downloads the parkings for a certain zone
* `deleteZone(...)`  that deletes the downloaded parkings that are within that zone

### Main changes:
* `Parking` now has a new attribute: `tile` which identifies to what tile it belongs
* `Tile` is now a type alias for `String`, this simplifies the management and creation of tiles, the "static" functions that were associated to it are now in a new object `TileUtils`.
* A new interface `OfflineParkingRepository` that extends `ParkingRepository`, it adds the API for dowloading / deleting zones
* `ParkingViewModel` has a new attribute of type `OfflineParkingRepository`

### Testing the database:
The Unit/Android tests cover the database by testing the `OfflineParkingRepositoryRoom`.
When the app is running, it is possible to inspect the database via the "App Inspector" as explained here:
https://developer.android.com/studio/inspect/database

## Code coverage:
- `com.github.se.cyrcle.model.parking.offline`: 100%
- `ParkingViewModel`: 86%
- Global:
image

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt
Comment (andrepeiry) [inline app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt:50]: *[TEMPORARY COMMENT WHILE ANDROID STUDIO IS SYNCING]*
Wouldn't it be cleaner to also destroy the database as well ? I feel like this can be some bug source if we don't delete the database between tests. I'm ready to believe that JUnit recreates the whole environment each time, but I'm not an expert on this subject.

Comment (matttsss) [reply app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt:50]: Mhhhh I'm not to sure. I think that because the `@Before` creates a new database, it ovewrites the database already created and held in that reference. Since those test databases are in memory, they will simply be deleted by the garbage collector?

File: app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:1]: For this file, I'm bothered with the following :
- Documentation is missing on most of the function
- Exception handling could be done at this level, as one of the two following, notably for the enum, in which the ordinal can be corrupt and bigger than the number of entries
  - change the function to the following scheme `deserializeOrDefault(ordinal, default) { try {deserialize} catch(E) {return default} }`
  - Throw a "proper" deserialization error instead of relying on the "IndexOutOfBound` when dealing with enums.

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:203]: Just a quick note (non-blocking), is that an explanation for an `if-else` statement should be written as follow
```
if (condition) {
    // Explanation
    ... code ...
} else {
    // Explanation
    ... code ...
}
```
Instead of above the branching, as this would better indicate the whole thing
```
// Absolute value of x
if (x < 0) {
    x =-x;
} else {
    x = x;
}
```

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:699]: I haven't tried myself, but could this have been obtain more easily with a `.filter` ?

File: app/src/main/java/com/github/se/cyrcle/model/parking/Tile.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/model/parking/Tile.kt:98]: The documentation for `getBottomLeftPoint` as well as `getTopRightPoint` could be improve such as
```
@param tile a valid tile to get the bottom left point from.
@return the bottom left point of the tile, null if tile is not valid.
```

File: app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepositoryFirestore.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepositoryFirestore.kt:80]: Please mark this with a `TODO` or another easy searchable string so that we can find them later ^^

File: app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepository.kt
Comment (andrepeiry) [inline app/src/main/java/com/github/se/cyrcle/model/parking/online/ParkingRepository.kt:1]: Thank you for adding the documentation on `getReportsForParking` and `addReport` <3

**Review by andrepeiry — CHANGES_REQUESTED** [review]
# Good code in general

Your code is readable, and even if the PR is quite hefty, it is not a pain to read.

## Some improvements

I'm a bit bothered by the `Adaptater`s

- Documentation is missing on most of the function
- Exception handling could be done at this level, as one of the two following, notably for the enum, in which the ordinal can be corrupt and bigger than the number of entries (also don't forget to add some tests on this)
   - change the function to the following scheme `deserializeOrDefault(ordinal, default) { try {deserialize} catch(E) {return default} }`
   - Throw a "proper" deserialization error instead of relying on the `IndexOutOfBound` when dealing with enums.

Also see some inline comments I left inside the code

File: app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt
Comment (LouisJamesVasseur) [reply app/src/androidTest/java/com/github/se/cyrcle/model/parking/OfflineParkingRepositoryRoomTest.kt:50]: Would you be willing to test both and report back? This could be a MAJOR error source so I'd be grateful if you tried before merging the PR. I am however not an expert either, so if you're confident I'll take your word for it

File: app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt
Comment (LouisJamesVasseur) [inline app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt:30]: any idea why it wasn't using the runWith before?

File: app/src/main/java/com/github/se/cyrcle/io/serializer/LocationAdapter.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/io/serializer/LocationAdapter.kt:1]: As @andrepeiry mentions, documentation is partly missing here

File: app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:42]: Yeah I definitely agree w/ @andrepeiry here. If we're dealing w/ indices in a table we need to error handle... it'd be quite a shame for the app to consistenly crash if we make a change at this level

File: app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:52]: same comment as above

File: app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/io/serializer/ParkingAdapter.kt:63]: here too :)

File: app/src/main/java/com/github/se/cyrcle/di/DatabaseModule.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/di/DatabaseModule.kt:15]: Maybe some doc here?

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:205]: Why is this here? This is "false" branching of the if condition so that means that tilesToParking[tile] is already an empty list, no?

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:699]: I guess? But since there is only one element to filter out, I'm not sure if it would be that useful...

File: app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt
Comment (matttsss) [reply app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt:30]: No, but it can only help to have them

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:704]: Again, a quick error handle to check that tilesToDelete > tilesToKeep would be nice

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:205]: Since requests to Firestore are asynchronous, if we have a second call to `getParkingsInRect`, it will make a second request to firestore if the first one did not complete before. This way the empty list marks this tile as "already" queried

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:704]: Since those are sets, the difference can only be the empty set in the worst case :)

File: app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt:1]: documentation here please :)

File: app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneSelectionScreen.kt
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/ui/zone/ZoneSelectionScreen.kt:121]: Should be in strings.xml

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:704]: Good point!

File: app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt
Comment (LouisJamesVasseur) [reply app/src/androidTest/java/com/github/se/cyrcle/ui/zone/ZoneManagerScreenTest.kt:30]: Agree with you there

**Comment by matttsss** [comment]
@andrepeiry @LouisJamesVasseur I hear your points on the serialization for the parking enums. However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you?
Also, I did not bother with documenting those `@TypeConverter` functions since it is quite straight forward what they do, but I could add that if necessary.

File: app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt
Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt:1]: It's inherited from the interfaces :)

**Review by LouisJamesVasseur — COMMENTED** [review]
VERY good work overall. The length of the PR definitely makes sense and I think you did well to put all of this into 1 PR. Some light implementation comments on my end but I think @andrepeiry said most of what I wanted to say already so I'll leave it at that. Mind the comments and I will approve!

File: app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/offline/OfflineParkingRepositoryRoom.kt:1]: Yeah okay, I see. No need then

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:205]: Good implementation then!

**Comment by LouisJamesVasseur** [comment]
> @andrepeiry @LouisJamesVasseur I hear your points on the serialization for the parking enums. However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you? Also, I did not bother with documenting those `@TypeConverter` functions since it is quite straight forward what they do, but I could add that if necessary.

The arbitrary enum solution works for me, I think what you're suggesting makes sense. Understood on the documentation, honestly it was more out of concern about the documentation aspect of M3 since you added a lot of code here. I'll leave it up to your opinion if you want to add the doc you can otherwise it's non-blocking on my end

File: app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt
Comment (LouisJamesVasseur) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:699]: I think the flatMap as is is appropriate

Comment (andrepeiry) [reply app/src/main/java/com/github/se/cyrcle/model/parking/ParkingViewModel.kt:699]: Well then, let this stay as is, it was merely a suggestion ^^

**Comment by andrepeiry** [comment]
> However, I do not use these functions, they are only used internally to Room and the function signature is quite limited. The only thing I could do is "if the index is out of bounds", return a arbitrary enum value if that doesn't bother you?

It's bothering me, as such exception could crash the app (I know it won't technically happens, as data corruption should be a quite negligible risk). Indeed, `if (corruption) return default` seems like a good enough solution for me ^^
If they are only used internally, maybe mark them as `private` ?

> Also, I did not bother with documenting those @TypeConverter functions since it is quite straight forward what they do, but I could add that if necessary.

I know documentation is as unloved as unit tests (for most human at least 😉), but it is always appreciated, and if I remember correctly, you are using GitHub Copilot, which is a great tool to autocomplete minimal documentation

**Comment by matttsss** [comment]
Thank you both,  I have addressed your comments if you want to check.
@andrepeiry Concerning marking the `@TypeConverters` as `private`, that will not be possible since `Room` needs to access them from outside.

**Review by LouisJamesVasseur — APPROVED** [review]
Thanks for addressing the changes I requested! Great work!

**Review by andrepeiry — APPROVED** [review]
Thank you for addressing our comments, this PR is good to go !

**Pull request merged**
